copycat/runtime/src/main/java/org/apache/kafka/copycat/runtime/distributed/KafkaConfigStorage.java

```diff
@@ -0,0 +1,438 @@
+        }
+
+        // In theory, there is only a single writer and we shouldn't need this lock since the background thread should
+        // not invoke any callbacks that would conflict, but in practice this guards against inconsistencies due to
+        // the root config being updated.
+        Map<String, Integer> newRootConfig = new HashMap<>();
+        synchronized (lock) {
+            // Validate tasks in this assignment. Any task configuration updates should include updates for *all* tasks
+            // in the connector -- we should have all task IDs 0 - N-1 within a connector if any task is included here
+            Map<String, Set<Integer>> updatedConfigIdsByConnector = new HashMap<>();
+            for (Map.Entry<ConnectorTaskId, Map<String, String>> taskConfigEntry : configs.entrySet()) {
+                ConnectorTaskId taskId = taskConfigEntry.getKey();
+                if (!updatedConfigIdsByConnector.containsKey(taskId.connector()))
+                    updatedConfigIdsByConnector.put(taskId.connector(), new TreeSet<Integer>());
+                updatedConfigIdsByConnector.get(taskId.connector()).add(taskId.task());
+            }
+            Map<String, Integer> updatedRootConfigEntries = new HashMap<>();
+            for (Map.Entry<String, Set<Integer>> taskConfigSetEntry : updatedConfigIdsByConnector.entrySet()) {
+                for (Integer elem : taskConfigSetEntry.getValue()) {
+                    if (elem < 0 || elem >= taskConfigSetEntry.getValue().size()) {
+                        log.error("Submitted task configuration contain invalid range of task IDs, ignoring this submission");
+                        throw new CopycatException("Error writing task configurations: found some connectors with invalid connectors");
+                    }
+                }
+                updatedRootConfigEntries.put(taskConfigSetEntry.getKey(), taskConfigSetEntry.getValue().size());
+            }
+
+            // Generate new root config
+            newRootConfig.putAll(rootConfig);
+            newRootConfig.putAll(updatedRootConfigEntries);
+
+            // Second half of validation: if we find some missing task configurations even with a) outstanding deferred
+            // updates and b) these updates, then we shouldn't generate an updated root config
+            boolean inconsistent = false;
+            for (Map.Entry<String, Integer> newRootEntry : newRootConfig.entrySet()) {
+                String connName = newRootEntry.getKey();
+                for (int taskIndex = 0; taskIndex < newRootEntry.getValue(); taskIndex++) {
+                    ConnectorTaskId taskId = new ConnectorTaskId(connName, taskIndex);
+                    // Note that we *do not* include deferred updates here. This is important because a half completed
+                    // write of updated task configs could leave some left over data in the deferred tasks, which would
```

Comment:
"Deferred tasks" would be the shorthand I used for deferred task configuration updates. No new concept there. Cleaned up the comment to clarify.